plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/index/HudiPartitionStatsIndexSupport.java

```diff
@@ -0,0 +1,102 @@
+
+    public HudiPartitionStatsIndexSupport(HoodieTableMetaClient metaClient)
+    {
+        super(log, metaClient);
+    }
+
+    @Override
+    public Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(HoodieTableMetadata metadataTable, Map<String, List<FileSlice>> inputFileSlices, TupleDomain<String> regularColumnPredicates)
+    {
+        return Map.of();
+    }
+
+    public Optional<List<String>> prunePartitions(HoodieTableMetadata metadataTable, TupleDomain<String> regularColumnPredicates)
+    {
+        // Filter out predicates containing simple null checks (`IS NULL` or `IS NOT NULL`)
+        TupleDomain<String> filteredRegularPredicates = regularColumnPredicates.filter((_, domain) -> !hasSimpleNullCheck(domain));
+
+        // Sanity check, if no regular domains, return immediately
+        if (filteredRegularPredicates.getDomains().isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<String> regularColumns = new ArrayList<>(filteredRegularPredicates.getDomains().get().keySet());
+
+        // Get columns to filter on
+        List<String> encodedTargetColumnNames = regularColumns
+                .stream()
+                .map(col -> new ColumnIndexID(col).asBase64EncodedString()).toList();
+
+        // Map of partition stats keyed by partition name
+        // If some columns in encodedTargetColumnNames is not available in partition stats index, partition will not be pruned
+        // This is so as we cannot be sure that the constraint is true or not, hence, adopt conservative measure
+        Map<String, List<HoodieMetadataColumnStats>> statsByPartitionName = metadataTable.getRecordsByKeyPrefixes(
+                        encodedTargetColumnNames,
+                        HoodieTableMetadataUtil.PARTITION_NAME_PARTITION_STATS, true)
+                .collectAsList()
+                .stream()
+                .filter(f -> f.getData().getColumnStatMetadata().isPresent())
+                .map(f -> f.getData().getColumnStatMetadata().get())
+                .collect(Collectors.groupingBy(HoodieMetadataColumnStats::getFileName));
+
+        List<String> prunedPartitions = statsByPartitionName.entrySet()
+                .stream()
+                .filter(e -> evaluateStatisticPredicate(filteredRegularPredicates, e.getValue(), regularColumns)).map(e -> e.getKey())
+                .toList();
+
+        return Optional.of(prunedPartitions);
+    }
+
+    public static boolean isIndexSupportAvailable(HoodieTableMetaClient metaClient)
+    {
+        return metaClient.getTableConfig().getMetadataPartitions().contains(HoodieTableMetadataUtil.PARTITION_NAME_PARTITION_STATS);
+    }
+
+    public static boolean shouldUseIndex(TupleDomain<String> tupleDomain, HoodieTableMetaClient metaClient)
+    {
+        if (!isIndexSupportAvailable(metaClient)) {
+            return false;
+        }
+        List<String> sourceFields = metaClient.getIndexMetadata().get().getIndexDefinitions()
```

Comment:
Unchecked Option.get() - let's check if index metadata is present or not? If not, then bail out.

Reply:
Fixed